nemo/src/execution/execution_engine.rs

```diff
@@ -378,4 +406,255 @@ impl<Strategy: RuleSelectionStrategy> ExecutionEngine<Strategy> {
+            ))));
+        }
+
+        // Rule index of the rule that was applied to derive the given fact
+        let rule_index = self.rule_history[step];
+        let rule = &self.program.rules()[rule_index];
+
+        let predicate_types: &Vec<PrimitiveType> = self
+            .analysis
+            .predicate_types
+            .get(&fact_predicate)
+            .expect("Every predicate should be associated with a type.");
+
+        // Iterate over all head atoms which could have derived the given fact
+        for (head_index, head_atom) in rule.head().iter().enumerate() {
+            if head_atom.predicate() != fact_predicate {
+                continue;
+            }
+
+            // Unify the head atom with the given fact
+            // If unification is possible `compatible` remain true
+            // and `assignment` will contain the match which is responsible for the fact
+            let mut compatible = true;
+            let mut assignment = VariableAssignment::new();
+
+            for (ty, (head_term, fact_term)) in predicate_types
+                .iter()
+                .zip(head_atom.terms().iter().zip(fact_terms.iter()))
+            {
+                if head_term.is_ground() {
+                    if ty.ground_term_to_data_value_t(head_term.clone())
+                        != ty.ground_term_to_data_value_t(fact_term.clone())
+                    {
+                        compatible = false;
+                        break;
+                    }
+                } else if let Term::Variable(variable) = head_term {
+                    // Matching with existential variables should not produce any restrictions,
+                    // so we just consider universal variables here
+                    if variable.is_existential() {
+                        continue;
+                    }
+
+                    if rule.constructors().contains_key(variable) {
+                        // TODO: Support arbitrary operations in the head
+                        return Err(Error::TraceUnsupportedFeature());
+                    }
+
+                    match assignment.entry(variable.clone()) {
+                        Entry::Occupied(entry) => {
+                            if ty.ground_term_to_data_value_t(entry.get().clone())
+                                != ty.ground_term_to_data_value_t(fact_term.clone())
+                            {
+                                compatible = false;
+                                break;
+                            }
+                        }
+                        Entry::Vacant(entry) => {
+                            entry.insert(fact_term.clone());
+                        }
+                    }
+                } else {
+                    unreachable!("Variables are the only non-ground terms");
+                }
+            }
+
+            if !compatible {
+                // Fact could not haven unified
```

Comment:
Not sure what this should say (`could not be unified`, maybe?), but `haven` is surely not the right word here.

Reply:
Had a little stroke there or something